### PR TITLE
ispurine and ispyrimidine for nucleotides

### PR DESCRIPTION
**Pull request by Ward9250**
Implements ispurine and ispyrimiadine for nucleotide types.

These methods return true if a nucleotide is surely a purine or is surely a pyrimidine. 
Where by "surely" we mean the base of the nucleotide is known with no ambiguity, or the ambiguity of the base does not change the certainty of whether the base is purine or pyrimadine.

### CODE REVIEW

**Comment by Ward9250** [comment]
This is done and RTR. @bicycle1885 Do you think we could increase performance for getting pyrimidines and purines in BioSequences by doing some clever bit ops on whole chunks of sequence at a time rather than getting each nucleotide and testing it individually? Signature would look something like `isPurine(seq::BioSequence)` and would return bit or logical vectors indicating which nucleotides are purine.

File: src/seq/nucleotide.jl [hunk 1]
```diff
@@ -103,6 +103,24 @@ function isGC(nt::DNANucleotide)
 end
 
 """
+    isPurine(nt::Nucleotide)
+
+Test if nucleotide is surely a Purine.
+"""
+function isPurine(nt::Nucleotide)
```
Comment (bicycle1885) [inline src/seq/nucleotide.jl]: camelCase is not the Julia naming convention. It should be `ispurine`.

File: src/seq/nucleotide.jl [hunk 2]
```diff
@@ -103,6 +103,24 @@ function isGC(nt::DNANucleotide)
...
+    return (compatbits(nt) | 0b0101) == 0b0101
```
Comment (bicycle1885) [inline src/seq/nucleotide.jl]: This is wrong. When `nt = DNA_Gap` `compatbits(nt)` is `0b0000` and hence `(compatbits(nt) | 0b0101) = 0b0101`. And the result is `isPurine(DNA_Gap)` is `true`.

File: src/seq/nucleotide.jl [hunk 3]
```diff
@@ -103,6 +103,24 @@ function isGC(nt::DNANucleotide)
...
+end
+
+"""
+    isPyrimidine(nt::Nucleotide)
+
+Test if nucleotide is surely a Pyrimidine.
+"""
+function isPyrimidine(nt::Nucleotide)
```
Comment (bicycle1885) [inline src/seq/nucleotide.jl]: `ispyrimidine`

File: src/seq/nucleotide.jl (hunk 2 above)
Comment (Ward9250) [inline src/seq/nucleotide.jl]: Can this be done with bits operations another way, or do we have to resort to a `nt == DNA_A || nt == DNA_G || nt == DNA_R`, and create seperate method defs out for DNANucleotide and RNANucleotide?

Comment (bicycle1885) [inline src/seq/nucleotide.jl]: Also note that `DNA_R`, an ambiguous DNA nucleotide, is surely a purine because R = A or G.

Comment (bicycle1885) [inline src/seq/nucleotide.jl]: Also note that `DNA_R`, an ambiguous DNA nucleotide, is surely a purine because R = A or G.

Comment (Ward9250) [inline src/seq/nucleotide.jl]: Yes this is what I was trying to achieve, A, G and R, should return true, all others should be false. I believe currently A G and R do return true, and the others do return false - except gaps which also return true.

Comment (bicycle1885) [inline src/seq/nucleotide.jl]: I think whay you intended is: `(compatbits(nt) & 0b0101) != 0x00`.

File: src/seq/Seq.jl [hunk 4]
```diff
@@ -46,6 +46,8 @@ export
     RNA_Gap,
     iscompatible,
     isambiguous,
+    isPurine,
+    isPyrimidine,
```
Comment (kdmurray91) [inline src/seq/Seq.jl]: ITYM `ispurine`? i.e. lower case P per the function definitions?

Comment (diegozea) [inline src/seq/Seq.jl]: I'm in favor if ispurine (lowercase P).

**Comment by Ward9250** [comment]
Awesome thanks guys, acted on all review issues, now waiting to see if tests pass.

File: test/seq/runtests.jl [hunk 5]
```diff
@@ -307,11 +307,29 @@ end
...
         end
         for x in alphabet(RNANucleotide)
-            @test isambiguous(x) == (x > RNA_U)
+            @test isambiguous(nt) == (nt > RNA_U)
+        end
+    end
+
+    @testset "isPurine" begin
+        for nt in alphabet(DNANucleotide)
+            @test isPurine(nt) == (nt == DNA_A || nt == DNA_G)
```
Comment (bicycle1885) [inline test/seq/runtests.jl]: You didn't update names in tests :(

Comment (Ward9250) [inline test/seq/runtests.jl]: Oh my this morning is not my morning!

File: src/seq/nucleotide.jl (hunk 2 above)
Comment (Ward9250) [inline src/seq/nucleotide.jl]: That still does not work, it stops gaps returning true, but DNA_M and others return true when only DNA_A, DNA_G, and DNA_R should return true, so currently any ambiguity code that could be A, and any code that could be G returns true. We only want, A, G, or A|G to return true.

Comment (bicycle1885) [inline src/seq/nucleotide.jl]: Sorry, I was careless. The following snippet should work:
```jl
function ispurine(nt::Nucleotide)
    bits = compatbits(nt)
    return ((bits & 0b1010) == 0) && ((bits & 0b0101) != 0)
end
```

**Pull request merged**